File: Client/src/lib/doppler.ts
export interface DopplerPlotData {
  time: number[]
  amplitude?: number[]
  frequency?: number[]
  stats?: DopplerStats
}

export interface DopplerParams {
  frequency: number
  velocity: number
  duration: number
}

export interface DopplerStats {
  max_observed?: number
  min_observed?: number
  shift_ratio?: number
}

export interface DopplerPrediction {
  predictedFrequency: number
  predictedVelocity: number
}

const BASE_URL = "http://127.0.0.1:8000/api/doppler"

/**
 * Fetch Doppler audio file from backend
 */
export async function fetchDopplerAudio(params: DopplerParams): Promise<Blob> {
  const url = `${BASE_URL}/audio?frequency=${params.frequency}&velocity=${params.velocity}&duration=${params.duration}`

  const response = await fetch(url, {
    method: "GET",
  })

  if (!response.ok) {
    throw new Error(`Failed to fetch audio: ${response.status} ${response.statusText}`)
  }

  return await response.blob()
}

/**
 * Fetch Doppler plot data (time, amplitude, frequency arrays) from backend
 */
export async function fetchDopplerPlotData(params: DopplerParams): Promise<DopplerPlotData> {
  const url = `${BASE_URL}/?frequency=${params.frequency}&velocity=${params.velocity}&duration=${params.duration}`

  const response = await fetch(url, {
    method: "GET",
  })

  if (!response.ok) {
    throw new Error(`Failed to fetch plot data: ${response.status} ${response.statusText}`)
  }

  const data = await response.json()
  return data as DopplerPlotData
}

/**
 * Predict source frequency and velocity from uploaded audio file using AI model
 */
export async function predictDoppler(audioFile: File): Promise<DopplerPrediction> {
  const formData = new FormData()
  formData.append("file", audioFile)

  const response = await fetch(`http://127.0.0.1:8000/api/dopplerAnalyze`, {
    method: "POST",
    body: formData,
  })

  if (!response.ok) {
    throw new Error(`Failed to predict: ${response.status} ${response.statusText}`)
  }

  const data = await response.json()
  return data as DopplerPrediction
}

File: Server/routers/doppler.py
from fastapi import APIRouter, Query
from fastapi.responses import StreamingResponse, JSONResponse
from services.doppler_service import simulate_doppler, get_first_predicted_speed
from fastapi import UploadFile, File, HTTPException
import shutil
import os
router = APIRouter()


@router.get("/doppler")
async def doppler(
    frequency: float = Query(..., description="Source frequency (Hz)"),
    velocity: float = Query(..., description="Source velocity (m/s)"),
    duration: float = Query(..., description="Duration (s)")
):
    """
    Simulate Doppler effect siren.
    Returns: audio (WAV) + JSON with time, amplitude, frequency.
    """
    try:
        result = simulate_doppler(frequency, velocity, duration)

        # Return both JSON (plot data) and WAV stream
        return {
                "time": result["time"],
                "amplitude": result["amplitude"],
                "frequency": result["frequency"],
                "stats":{
                    "max_observed": result["max_frequency"],
                    "min_observed": result["min_frequency"],
                    "shift_ratio": result["shift_ratio"]
                }
            }

    except Exception as e:
        return JSONResponse(content={"error": str(e)}, status_code=500)

@router.get("/doppler/audio")
async def doppler_audio(
    frequency: float = Query(...),
    velocity: float = Query(...),
    duration: float = Query(...),
):
    """Stream Doppler audio only (WAV)"""
    result = simulate_doppler(frequency, velocity, duration)
    return StreamingResponse(result["audio"], media_type="audio/wav")




UPLOAD_DIR = "uploads"  # directory to temporarily save uploaded audio files
os.makedirs(UPLOAD_DIR, exist_ok=True)

@router.post("/dopplerAnalyze")
async def doppler_analyze(file: UploadFile = File(...)):
    """
    Upload an audio file named like 'VehicleName_trueSpeed.wav'
    and get the first predicted speed from the hardcoded model.
    """
    try:
        # 1️⃣ Save the uploaded file temporarily
        file_path = os.path.join(UPLOAD_DIR, file.filename)
        with open(file_path, "wb") as buffer:
            shutil.copyfileobj(file.file, buffer)

        # 2️⃣ Process the file
        result = get_first_predicted_speed(file_path)

        # 3️⃣ Optionally, clean up the uploaded file
        os.remove(file_path)

        return {"predictedVelocity": result, "predictedFrequency": 440}  # Placeholder frequency

    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))
